### ISSUE TITLE
D103 vs pylint counterpart C0116 (undocumented-public-function)

### ISSUE THREAD

**Issue by sigma67** [astral-sh/ruff#9149]
As noted in https://github.com/astral-sh/ruff/issues/970#issuecomment-1341901539 , in #970 D103 is different from its pylint counterpart C0116.

 It only requires that a function has docstring some in class hierarchy and does not require over-ridden methods to have a docstring.

``` python
class Foo:
  def validate(self):
    """Doc."""
    ...

class Bar(Foo):
  def validate(self):
    ...
```

I'm creating a separate issue to give this topic some visibility and to discuss it.

**Comment by sigma67** [comment]
The same applies to `D105` when overriding a base class' magic method, which already has a docstring. In pylint `C0116` this is ok, in `D105` it's not.

**Comment by sigma67** [comment]
@charliermarsh Perhaps you can comment if this is intended or could be changed? This would not be a breaking change, more of a "softening" of the rule to be more compatible with repositories moving to ruff from pylint.

CC @hmc-cs-mdrissi original report

**Comment by charliermarsh** [comment]
@sigma67 - The challenge here is that, unlike Pylint, Ruff doesn't do analysis _across_ files right now. So we could support this _within_ a file, but we wouldn't be able to detect that the `validate` on `Bar` is an implementation of a parent interface if `Foo` is defined in another file.

One valid solution here is that we do ignore `D102` if you mark the method with `typing.override` or `typing_extensions.override`, which signals to type checkers that the method is intended to override a parent method.

**Comment by sigma67** [comment]
@charliermarsh  I see, thanks for the clarification and the hint regarding `typing.override`, that's quite helpful.

Are there future plans for cross-file analysis?

**Comment by charliermarsh** [comment]
@sigma67 -- Yes, it will absolutely be part of Ruff in the future. It's a big project though so will take some time :)
